File: src/test/scala/akkastudy/becomeactor/scala/BecomeActorTest.scala
package akkastudy.becomeactor.scala

import org.scalatest.FlatSpec
import org.scalatest.matchers.MustMatchers
import akka.actor.{ActorSystem, Props}
import akkastudy.SenderActor

class BecomeActorTest extends FlatSpec with MustMatchers {
  behavior of "An Actor System should"

  it should "An actor can rotate state with become and unbecome" in {
    val system = ActorSystem("MySystem")
    val becomeActor = system.actorOf(Props[BecomeActor], name = "becomeActor")
    implicit val senderActor =
      system.actorOf(Props[SenderActor], name = "senderActor") //the implicit actor that is sent
    (becomeActor!("goHiking"))
    (becomeActor!("goHiking"))
    (becomeActor!("useFoulLanguage"))
    (becomeActor!("goHiking"))              //No
    (becomeActor!("getDrunkAndCrashCar"))   //No
    (becomeActor!("goHiking"))              //No
    (becomeActor!("askForgiveness"))        //No
    (becomeActor!("spendTimeInPrison"))     //No
    (becomeActor!("goOutToDinner"))         //Nom
    (becomeActor!("askForgiveness"))        //No Response
    (becomeActor!("goHiking"))              //Good Hike!
  }
}

File: src/test/java/akkastudy/becomeactor/java/BecomeActorTest.java
package akkastudy.becomeactor.java;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akkastudy.remote.java.SenderActorJava;
import org.junit.Test;

/**
 * @author Daniel Hinojosa
 * @since 3/24/14 3:35 PM
 * url: <a href="http://www.evolutionnext.com">http://www.evolutionnext.com</a>
 * email: <a href="mailto:dhinojosa@evolutionnext.com">dhinojosa@evolutionnext.com</a>
 * tel: 505.363.5832
 */
public class BecomeActorTest {
    @Test
    public void testBasicAsynchronousAsk() throws InterruptedException {
        ActorSystem system = ActorSystem.create("MySystem");
        ActorRef becomeActor = system.actorOf(Props.create(BecomeActor.class), "becomeActor");
        ActorRef senderActor = system.actorOf(Props.create(SenderActorJava.class), "senderActor");

        becomeActor.tell("goHiking", senderActor);
        becomeActor.tell("goHiking", senderActor);
        becomeActor.tell("useFoulLanguage", senderActor);
        becomeActor.tell("goHiking", senderActor); //No
        becomeActor.tell("getDrunkAndCrashCar", senderActor); //No
        becomeActor.tell("goHiking", senderActor); //No
        becomeActor.tell("askForgiveness", senderActor); //No
        becomeActor.tell("spendTimeInPrison", senderActor); //No
        becomeActor.tell("goOutToDinner", senderActor); //Nom
        becomeActor.tell("askForgiveness", senderActor); //No Response
        becomeActor.tell("goHiking", senderActor); //Good Hike!
    }
}
